### PR TITLE
README for Airlift

### PR DESCRIPTION
**Pull request by schrockn**
## Summary & Motivation

Updated, more grounded Airlift README with code examples, etc.

v5: https://github.com/dagster-io/dagster/blob/4a5ae1fea216d141e9f8462c84412f7fc4d72bad/examples/experimental/dagster-airlift/README.md

## How I Tested These Changes

Read

### CODE REVIEW

**Comment by schrockn** [comment]
* **#23533** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23533?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @schrockn and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by schrockn — COMMENTED** [review]
Other option is to comment on hackmd where I authored: https://hackmd.io/7lYledE_SPCwwNzszuoApA

File: examples/experimental/dagster-airlift/README.md [hunk 1]
```diff
@@ -1,26 +1,174 @@
...
-The goal of Airlift is to accelerate, lower the cost, and reduce the risk of migrating to Dagster from Airflow. The tool allows you to:
+Airlift is a toolkit for observing Airflow instances from within Dagster and for accelerating the migration of Airflow DAGs to Dagster assets.
 
-- Visualize and observe your existing Airflow DAGs within Dagster with no code changes in Airflow and few lines of code in your Dagster deployment.
-- Model the assets produced by your Airflow instance prior to migration in Dagster (e.g. make your dbt models appear in Dagster while they are orchestrated by Airflow).
-- Enable incremental migration of your Airflow tasks one-by-one, in any order. While the migration is happening, your Airflow DAG structures remain untouched from the standpoint of the observer.
+## Goals
 
-Tl;dr With Airlift, Dagster first becomes your “single pane of glass” for all orchestration and then acts as a toolkit to accelerate a low-risk migration process. Another design partner we are talking to describes this as “changing the oil while driving the car” which I think is a memorable analogy.
+- Observe Airflow DAGs and their execution history with no changes to Airflow code
```
Comment (sryza) [inline examples/experimental/dagster-airlift/README.md:7]: It might be worth adding something like "in order to trigger work in Dagster after Airflow DAGs complete"

File: examples/experimental/dagster-airlift/README.md [hunk 2]
```diff
@@ -1,26 +1,174 @@
...
+- **Astronomer:** No limitations as far as we know.
 
-## Trying out airlift
+## Peering
 
-See the peering process in action by trying out the example:
+The first step is to peer the Dagster Deployment and the Airflow instance.
 
-- [Airflow-Dagster Peering with DBT](examples/peering-with-dbt)
+To do this you need to install airlift in and make it available in your Dagster deployment.
```
Comment (sryza) [inline examples/experimental/dagster-airlift/README.md:38]:
```suggestion
To do this you need to install airlift and make it available in your Dagster deployment.
```

File: examples/experimental/dagster-airlift/README.md [hunk 3]
```diff
@@ -1,26 +1,174 @@
...
+An MWAA auth backend is available at `dagster_airlift.mwaa.auth.MwaaSessionAuthBackend`
+
+This function creates:
+
+- An asset representing each DAG. This asset is “materialized” whenever a DAG run completes.
+- A sensor that polls the Airflow instance for operational information. This is what keeps the DAG execution history up to date. It must be on to get timely information.
+
+### Peering to multiple instances
+
+Airlift supports peering to multiple Airflow instances, as you can invoke `create_airflow_instance_defs` multiple times and combine them with `Definitions.merge`
```
Comment (sryza) [inline examples/experimental/dagster-airlift/README.md:72]:
```suggestion
Airlift supports peering to multiple Airflow instances, as you can invoke `create_airflow_instance_defs` multiple times and combine them with `Definitions.merge`:
```

File: examples/experimental/dagster-airlift/README.md (hunk 1 above)
Comment (schrockn) [reply examples/experimental/dagster-airlift/README.md:7]: I think that will be confusing because that implies that is the only reason why you would use this feature.

Comment (sryza) [reply examples/experimental/dagster-airlift/README.md:7]: Yeah agree that it doesn't make sense to frame it as the only reason to have observability.

Main comment is that I had thought that was one of the main goals and was surprised not to see it mentioned here.

File: examples/experimental/dagster-airlift/README.md [hunk 4]
```diff
@@ -1,26 +1,174 @@
...
+## REST API Availability
 
-Dagster is the “source of truth” for the state of migration as you can tell what assets and tasks have been fully migrated and which ones have not.
+Airlift depends on the the availability of Airflow’s REST API. Airflow’s REST API was made stable in its 2.0 release (Dec 2020) and was introduced experimentally in 1.10 in August 2018. Currently Airflow requires the availability of the REST API.
 
-This will greatly accelerate the migration because you can migrate tasks in parallel, in any order, with high trust, and do so aggressively because rolling back is easy, all within an orderly, observed, tool-assisted process.
+- **OSS:** Stable as of 2.00
+- **MWAA**
+  - Note: only available in Airflow 2.4.3 or later on MWAA.
+  - Doesn’t appear to be a well-defined reason other than pure conservatism according to [ChatGPT](https://chatgpt.com/c/220cd63e-2111-4b32-bd83-87d95e2390d7).
```
Comment (dpeng817) [inline examples/experimental/dagster-airlift/README.md:30]: probably remove this line?

File: examples/experimental/dagster-airlift/README.md [hunk 5]
```diff
@@ -1,26 +1,174 @@
...
-## Trying out airlift
+## Peering
 
-See the peering process in action by trying out the example:
+The first step is to peer the Dagster Deployment and the Airflow instance.
 
-- [Airflow-Dagster Peering with DBT](examples/peering-with-dbt)
+To do this you need to install airlift in and make it available in your Dagster deployment.
+
+```bash
```
Comment (dpeng817) [inline examples/experimental/dagster-airlift/README.md:40]: pip install uv as first step

File: examples/experimental/dagster-airlift/README.md [hunk 6]
```diff
@@ -1,26 +1,174 @@
...
+```
+
+At that point you create a `Definitions` object using `build_defs_from_airflow_instance`.
+
+```python
+from dagster_airlift.core import AirflowInstance, build_defs_from_airflow_instance
+
+defs = build_defs_from_airflow_instance(
+    airflow_instance=AirflowInstance(
+        auth_backend=BasicAuthBackend(
```
Comment (dpeng817) [inline examples/experimental/dagster-airlift/README.md:51]: comment here about how there's also other backends available. Namely, MWAA backend

File: examples/experimental/dagster-airlift/README.md [hunk 7]
```diff
@@ -1,26 +1,174 @@
...
+- A sensor that polls the Airflow instance for operational information. This is what keeps the DAG execution history up to date. It must be on to get timely information.
+
+### Peering to multiple instances
+
+Airlift supports peering to multiple Airflow instances, as you can invoke `create_airflow_instance_defs` multiple times and combine them with `Definitions.merge`
+
+```python
+from dagster_airlift.core import AirflowInstance, build_defs_from_airflow_instance
+
+defs = build_defs_from_airflow_instance(
```
Comment (dpeng817) [inline examples/experimental/dagster-airlift/README.md:77]: This isn't how definitions.merge works IIRC. It's actually `Definitions.merge(d1, d2)` (fully static).

I've been burned by this in the past

File: examples/experimental/dagster-airlift/README.md [hunk 8]
```diff
@@ -1,26 +1,174 @@
...
+To add definitions that an Airlift-enabled deployment will observed, you need to use the `orchestrated_defs` argument to `build_defs_from_airflow_instance`
+
+_Note: This also accepts an object of type `Definitions`. We have recently added a function, `Definitions.merge`, which allows users to combine and compose `Definitions` objects, which this utilizes._
+
+For most projects we anticipate the following process:
+
+- Create a new factory class for every operator type or specialized use case of an operator (e.g. a BashOperator invoking a dbt project).
+- Use that factory to create definitions associating with each Airflow task.
+
+In our example we have a pre-written factory class (in `examples/experimental/dagster-airlift/dagster_airlift/dbt/multi_asset.py`) that is associated with invoking a dbt project:
```
Comment (dpeng817) [inline examples/experimental/dagster-airlift/README.md:115]: mark that it is installable via the dbt submodule

File: examples/experimental/dagster-airlift/README.md [hunk 9]
```diff
@@ -1,26 +1,174 @@
...
+)
+```
+
+### Mapping assets to tasks
+
+We default to a "convention over configuration" approach to affiliate Dagster assets with Airflow tasks.
+
+Note the naming convention `dbt_dag__build_dbt_models`. By default we use this convention to encode the dag name (`dbt_dag`) and task id (`build_dbt_models`) of the Airflow task actually orchestrating the computation.
+
+This name gets parsed and set corresponding tags (`airlift/dag_id` and `airlift/task_id`). Alternatively you can set those tags explicitly if you don’t want to rely on the naming convention.
```
Comment (dpeng817) [inline examples/experimental/dagster-airlift/README.md:170]: Maybe worth noting that these tags are set on both the asset and the underlying nodedef, for people who know what that is.

Comment (dpeng817) [reply examples/experimental/dagster-airlift/README.md:170]: Also noting that you have to specify the tags under op_tags on an asset.

File: examples/experimental/dagster-airlift/README.md [hunk 10]
```diff
@@ -1,26 +1,174 @@
...
+
+Once this is done you should be able to reload your definitions and the see the full dbt project. Once you run the corresponding Airflow Dag and task, each task completion will corresponding to an asset materialization in any asset that is orchestrated by that task.
+
+_Note: There will be some delay as this process is managed by a Dagster sensor that polls the Airflow instance for task history. This is every 30 seconds by default, so there will be some delay._
```
Comment (dpeng817) [inline examples/experimental/dagster-airlift/README.md:174]: I've actually made the default sensor polling interval very aggressive for the demos. Currently like 1 second.
But likely still will be some delay.

**Review by dpeng817 — CHANGES_REQUESTED** [review]
Comments

File: examples/experimental/dagster-airlift/README.md (hunk 1 above)
Comment (schrockn) [reply examples/experimental/dagster-airlift/README.md:7]: I'm going to hold off on that until we add a code example of doing that. I think it should be its own section.

**Comment by schrockn** [comment]
Updated! Thanks for great feedback: https://github.com/dagster-io/dagster/blob/4a5ae1fea216d141e9f8462c84412f7fc4d72bad/examples/experimental/dagster-airlift/README.md

File: examples/experimental/dagster-airlift/README.md [hunk 11]
```diff
@@ -1,26 +1,178 @@
...
+
+For most projects we anticipate the following process:
+
+- Create a new factory class for every operator type or specialized use case of an operator (e.g. a BashOperator invoking a dbt project).
+- Use that factory to create definitions associating with each Airflow task.
+
+In our example we have a pre-written factory class (in `examples/experimental/dagster-airlift/dagster_airlift/dbt/multi_asset.py` and installable by `uv pip install dagster-airlift[dbt]`) that is associated with invoking a dbt project:
+
+```python
+@dataclass
```
Comment (dpeng817) [inline examples/experimental/dagster-airlift/README.md:122]: add import of `dagster_airlift.core.DefsFactory` (also worth ensuring that's exported top level)

**Review by dpeng817 — APPROVED** [review]
one more comment but lgtm

**Comment by schrockn** [comment]
### Merge activity

* **Aug 9, 6:22 PM EDT**: @schrockn started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/23533).
